// eslint-plugin-wideband/src/rules/wideband-internal-underscore.js
"use strict";

module.exports =
{
    meta:
    {
        type: "suggestion",
        docs:
        {
            description: "Ensure '_' identifier is only used for internal local variables.",
            recommended: false,
        },
        schema: [ ],
    },

    create ( context )
    {
        const _sourceCode = context.sourceCode;

        /**
         * Returns true if the given node is in the top-level scope (module or script root).
         */
        function _isTopLevel ( node )
        {
            const _scope = _sourceCode.getScope ( node );

            return _scope.type === "module" || _scope.type === "global";
        }

        /**
         * Handle variable declarations like: const _ = ...
         */
        function checkVariableDeclarator ( node )
        {
            if ( node.id?.type !== "Identifier" ) return;

            if ( node.id.name !== "_" ) return;

            if ( _isTopLevel ( node ) )
            {
                context.report (
                {
                    node: node.id,
                    message: "Identifier '_' should be reserved for internal local variables, not top-level declarations.",
                } );
            }
        }

        /**
         * Handle parameters named `_` only if the function is top-level.
         */
        function checkFunctionParams ( node )
        {
            if ( ! _isTopLevel ( node ) ) return;

            for ( const _param of node.params )
            {
                if ( _param.type === "Identifier" && _param.name === "_" )
                {
                    context.report (
                    {
                        node: _param,
                        message: "Top-level function parameter '_' is not allowed; use it only for internal locals.",
                    } );
                }
            }
        }

        return {
            VariableDeclarator:      checkVariableDeclarator,
            FunctionDeclaration:     checkFunctionParams,
            FunctionExpression:      checkFunctionParams,
            ArrowFunctionExpression: checkFunctionParams,
        };
    },
};
